### ISSUE TITLE
Unexpected schema for Record with Literal Union key

### ISSUE THREAD

**Issue by shwao** [sinclairzx81/typebox#176]
The Type.Record function generates a (for me) unexpected schema. The TypeScript type `Record<"one" | "two", string>` requires the object to have both, `one` and `two` properties. The schema however allows for empty objects or invalid keys.

```ts
import { Type } from "@sinclair/typebox";
import Ajv from "ajv";

const ajv = new Ajv({ keywords: ["kind"] });

const KEYS = ["one", "two"] as const;

const U = Type.Union(KEYS.map(key => Type.Literal(key)));

const R = Type.Record(U, Type.String());

const uValidator = ajv.compile(U);
const rValidator = ajv.compile(R);

/* I expect false since its not "one" or "two" */
console.log(uValidator("not one or two")); // false
/* I would expect false since it does not have "one" or "two" properties, but its true. */
console.log(rValidator({})); // true
/* I would expect this to be false and it is. */
console.log(rValidator({ "invalid key": "invalid key" })); // false
/* I would expect this to be false but its true. */
console.log(rValidator({ "oneee": "invalid key" })); // true

/* TypeScript does it differently */

type TypeScriptU = "one" | "two";
type TypeScriptR = Record<TypeScriptU, string>;

/* Type '{}' is missing the following properties from type 'TypeScriptR': one, two ts(2739) */
const invalidObject: TypeScriptR = {};

/*
	I can achieve my desired schema like this but this way its not generating the
	correct or desired TypeScript type, ie. not preserving the key literals:

	const WorkingR: TIntersect<TObject<{
			[x: string]: TString;
	}>[]>

	instead of:

	TRecord<TUnion<TLiteral<"one" | "two">[]>, TString>

	for the R schema.
*/
const WorkingR = Type.Intersect(
	KEYS.map(key => Type.Object({ [key]: Type.String() }))
);
const workingRValidator = ajv.compile(WorkingR);

console.log(workingRValidator({})); // false
console.log(workingRValidator({ "invalid key": "invalid key" })); // false
console.log(workingRValidator({ "oneee": "invalid key" })); // false
console.log(workingRValidator({ "one": "invalid because of missing two" })); // false
console.log(workingRValidator({ "one": "valid", "two": "valid" })); // true
```

Of course i could just create a `Type.Object({ "one": Type.String(), "two": Type.String() })` but the point is that i got this array that i want to use as keys.

Am i doing or expecting something wrong? I thought the TypeBox Record util should behave like the TypeScript Record util.

**Comment by sinclairzx81** [comment]
@shwao Hi, Apologies for the delay in response. 

The `Type.Record(...)` type tries mirror the TypeScript `Record<Key, Value>` utility type as best as possible under the rules of JSON schema. For `string` and `number` keys, the rules for validation should be equivalent, however for union string keys, you're right in that there is a disparity here in that `Type.Record()` doesn't behave exactly as a `Type.Object()`. This reason for this is `Type.Record()` is essentially an alias for JSON schema `patternProperties` and thus inherits it's validation rules (which are permissive)

Example

```typescript

// --------------------------------------------------
// Types
// --------------------------------------------------

type T = Record<'one' | 'two', string>

const T = Type.Record(Type.Union([
    Type.Literal('one'),
    Type.Literal('two')
]), Type.String(), {
    additionalProperties: false // important
})

// --------------------------------------------------
// Schema
// --------------------------------------------------

check(T, { })                                // schema: ok,   ts: fail
check(T, { one: '' })                        // schema: ok,   ts: fail
check(T, { two: '' })                        // schema: ok,   ts: fail
check(T, { one: '', two: '' })               // schema: ok,   ts: ok
check(T, { one: '', two: '', three: '' })    // schema: fail, ts: fail

// --------------------------------------------------
// TypeScript
// --------------------------------------------------

const T0: T = { }                             // schema: ok,   ts: fail
const T1: T = { one: '' }                     // schema: ok,   ts: fail
const T2: T = { two: '' }                     // schema: ok,   ts: fail
const T3: T = { one: '', two: '' }            // schema: ok,   ts: ok
const T4: T = { one: '', two: '', three: '' } // schema: fail, ts: fail
```

As you mentioned though, the expectation may be that `Type.Record(TUnion<[TLiteral]>)` should return something equivalent to `Type.Object({ one: Type.String(), two: Type.String() })` as per your example. But for this to be resolved, `Type.Record()` would need to omit `patternProperties` specifically for the string union case and just emit `properties` as a standard `Type.Object()`. I am a little bit apprehensive about implementing that as the return structure / schema for `Type.Record()` becomes a little ambiguous (implicating user code that may reflect on the structure of the `Type.Record()` with the expectation that `patternProperties` exists), but maybe there is a schema representation that might fit under the `patternProperties` (noting that computed properties through string template literal types may be considered in future releases).

TypeBox is due for a major release, so may spend some time looking into this. But would certainly be happy to discuss this in more depth. 

S

**Comment by sinclairzx81** [comment]
@shwao Have just released TypeBox `0.24.0` and the there new functionality there to try and align with TypeScript's interpretation of Records with known keys. The following implementation is now possible.

```typescript
const K = Type.Union([
    Type.Literal('A'),
    Type.Literal('B'),
    Type.Literal('C'),
])

const R = Type.Record(K, Type.String(), { 
    additionalProperties: false 
})
```
This is achieved by `Type.Record()` return objects of type `TObject` if the `TRecord` is initialized with string literal values. Note that in order to constrain to only these known values, you must specify `{ additionalProperties: false }` which inline with JSON schemas property checks.

Will close off this issue, but feel free to re-open if you run into any problems
Cheers
S

**Comment by shwao** [comment]
Thank you very much! It works now as i expected it.
